### ISSUE TITLE
Coscheduling: when some pod failed, the other pods don't exit

### ISSUE THREAD

**Issue by zhiyxu** [kubernetes-sigs/scheduler-plugins#369]
I'm pre-researching coscheduling plugin, combind with pytorch operator, trying to use it in production environment and replacing volcano scheduler, to avoid fragmentation of resource view, but I got some problems.

After all master and workers running successfully, if some worker killed or failed, in volcano, all pods under the job are restarted, which is expected. But when using scheduling plugin, if some worker failed, the others just keep running, which is meaningless in our situation, so I'm wondering if there is any suggestion to solve this problem.

**Comment by zhiyxu** [comment]
> 

@Huang-Wei Thank you for your quick reply.

In a distributed training task, if some child task failed, we expect the whole task exit or restarted, currently this is implemented by volcano controller using [following config](https://github.com/volcano-sh/volcano/blob/master/example/integrations/tensorflow/dist-mnist/tf-dist-mnist-example.yaml#L12):
```
  policies:
    - event: PodEvicted
      action: RestartJob
    - event: PodFailed
      action: RestartJob
```
Do we need to implement a controller to achieve this when using scheduler plugin, or is there any better solution?

**Comment by Huang-Wei** [comment]
> Do we need to implement a controller to achieve this when using scheduler plugin, or is there any better solution?

Yes, how to act on a failure of pods should be the responsibility of the job controller, instead of scheduler.

Regarding the controller implementation, does a _pure_ volcano controller fit? If no, does kubeflow training operator define a similar job-level restarting API to fit your req?

**Comment by denkensk** [comment]
> After all master and workers running successfully, if some worker killed or failed, in volcano, all pods under the job are restarted, which is expected. 

These behavior is need to provided by the controller instead of scheduler. Actually it's also provided in kubeflow/trainning-operator.  
Can you try to set `RestartPolicy` as `OnFailure ` in your PytrochJob.
https://github.com/kubeflow/common/blob/26a6cece18dde58174b493fe86cf0eb421b5e153/pkg/apis/common/v1/types.go#L156
@zhiyxu

**Comment by zhiyxu** [comment]
> > After all master and workers running successfully, if some worker killed or failed, in volcano, all pods under the job are restarted, which is expected.
> 
> These behavior is need to be provided by the controller instead of scheduler. Actually it's also provided in kubeflow/trainning-operator. Can you try to set `RestartPolicy` as `OnFailure ` in your PytrochJob. https://github.com/kubeflow/common/blob/26a6cece18dde58174b493fe86cf0eb421b5e153/pkg/apis/common/v1/types.go#L156 @zhiyxu

@denkensk Thanks, I'm testing PytorchJob workload, I have one more question:

Currently, we are using a mix of default-scheduler and volcano-scheduler, and scheduling single-machine and distributed offline tasks in different resource pools through `metadata.labels`, which results in the split-up of resource views and makes it difficult to improve the efficiency of resource usage. So we want to unify the scheduling view through scheduler-plugins, and replace volcano-scheduler which provide gang-scheduling capability by Coscheduling. 

I'd like to know if there is any experience for the migration from other teams, and what's the RoadMap of the scheduler-plugins project, thanks in advance.

**Comment by denkensk** [comment]
> I'd like to know if there is any experience for the migration from other teams.

We test the same jobs between volcano-scheduler and Coscheduling in scheduler-plugins. I don't think there is any extra work. Also there are some experience to share with you from the production like openAI https://openai.com/blog/scaling-kubernetes-to-7500-nodes/ and our company https://help.aliyun.com/document_detail/178169.html

**Comment by denkensk** [comment]
> what's the RoadMap of the scheduler-plugins project, thanks in advance.

Wei and me are proposing a refined PodGroup API in https://docs.google.com/document/d/16yLngEOd6x3IS6ejkclrCdmkDYfxtTyr97a7uNug_wA/view, and pursuing to make it available in k/k.

Fell free to give your comments.

**Comment by Huang-Wei** [comment]
Regarding this the roadmap of this repo, it will focus on the scheduling semantics and implementation, based on scheduler framework.
